src/wp-includes/class-wp-theme-json.php

```diff
@@ -2703,17 +2703,13 @@ private static function update_separator_declarations( $declarations ) {
 	 * @return array The block nodes in theme.json.
 	 */
 	private static function get_block_nodes( $theme_json, $selectors = array(), $options = array() ) {
-		$selectors = empty( $selectors ) ? static::get_blocks_metadata() : $selectors;
-		$nodes     = array();
-		if ( ! isset( $theme_json['styles'] ) ) {
-			return $nodes;
-		}
+		$nodes = array();
 
-		// Blocks.
 		if ( ! isset( $theme_json['styles']['blocks'] ) ) {
 			return $nodes;
 		}
 
+		$selectors               = empty( $selectors ) ? static::get_blocks_metadata() : $selectors;
 		$include_variations      = $options['include_block_style_variations'] ?? false;
 		$include_node_paths_only = $options['include_node_paths_only'] ?? false;
```

Comment:
```suggestion
		$include_variations      = $options['include_block_style_variations'] ?? false;
		$include_node_paths_only = $options['include_node_paths_only'] ?? false;

		if ( ! $include_node_paths_only ) {
			$selectors = empty( $selectors ) ? static::get_blocks_metadata() : $selectors;
		}
```

The `$selectors` only needed if we need full node list, for `$include_node_paths_only` we don't need it so we can remove it.

TT4 XHG benchmark shows it remove some extra calls for `get_blocks_metadata()`

<img width="1715" alt="Screenshot 2024-10-18 at 10 37 28 AM" src="https://github.com/user-attachments/assets/48410515-123a-49ba-847c-d27266008574">

Reply:
Nice!

Reply:
I'll update this PR and backport to Gutenberg. Thanks again, for testing.